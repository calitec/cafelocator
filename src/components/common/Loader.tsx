import { css } from '@emotion/react'
import media from '../../lib/styles/media'
import Spin from './Spin'
import Skeleton from './Skeleton'

const Loader: React.FunctionComponent = () => {
  return (
    <div css={LoaderWrapper}>
      <ul className="infoList">
        <Spin />
        <Skeleton />
        <Skeleton />
        <Skeleton />
      </ul>
    </div>
  )
}

const LoaderWrapper = css`
  display: block;
  width: 350px;
  height: 400px;
  overflow-y: auto;
  z-index: 1000;
  ${media.large} {
    width: 100%;
    height: 350px;
  }
`

export default Loader
