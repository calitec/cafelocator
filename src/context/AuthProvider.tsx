import * as React from 'react'
import { useEffect } from 'react'
import { useCallback } from 'react'
import { createContext, useState, useContext } from 'react'
import {
  getAuth,
  signOut,
  setPersistence,
  browserSessionPersistence,
  signInWithRedirect,
  onAuthStateChanged,
} from 'firebase/auth'
import { provider } from '../firebase'

interface State {
  user: any
  setUser: (data: object) => void
  onLogin: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void
  onLogout: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void
}

const AuthStateContext = createContext<State | null>(null)

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(null)
  const auth = getAuth()

  useEffect(() => {
    onAuthStateChanged(auth, (user) => {
      if (!user) return
      setUser({ ...user, user: user })
    })
  }, [])

  const onLogin = useCallback((e) => {
    e.preventDefault()
    try {
      setPersistence(auth, browserSessionPersistence)
        .then(() => {
          // const provider = new GoogleAuthProvider()
          // In memory persistence will be applied to the signed in Google user
          // even though the persistence was set to 'none' and a page redirect
          // occurred.
          return signInWithRedirect(auth, provider)
        })
        .catch((error) => {
          // Handle Errors here.
          const errorCode = error.code
          const errorMessage = error.message
        })

      // signInWithPopup(auth, provider)
      //   .then((result) => {
      //     // This gives you a Google Access Token. You can use it to access the Google API.
      //     const credential = GoogleAuthProvider.credentialFromResult(result)
      //     const token = credential.accessToken
      //     // The signed-in user info.
      //     const user = result.user
      //     setUser({ ...user, user: user })
      //     // ...
      //   })
      //   .catch((error) => {
      //     // Handle Errors here.
      //     const errorCode = error.code
      //     const errorMessage = error.message
      //     // The email of the user's account used.
      //     const email = error.email
      //     // The AuthCredential type that was used.
      //     const credential = GoogleAuthProvider.credentialFromError(error)
      //     // ...
      //   })
    } catch (error) {
      console.error(error)
    }
  }, [])

  const onLogout = useCallback((e) => {
    try {
      e.preventDefault()
      if (window.confirm('로그아웃 하시겠습니까?')) {
        signOut(auth)
          .then(() => {
            // Sign-out successful.
            setUser(null)
          })
          .catch((error) => {
            // An error happened.
          })
      }
    } catch (error) {
      console.error(error)
    }
  }, [])

  return (
    <AuthStateContext.Provider value={{ user, setUser, onLogin, onLogout }}>
      {children}
    </AuthStateContext.Provider>
  )
}

export function useAuthState() {
  const state = useContext(AuthStateContext)
  if (!state) throw new Error('Cannot find AuthStateContext')
  return state
}
