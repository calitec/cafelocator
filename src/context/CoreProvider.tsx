import * as React from 'react'
import { createContext, useState, useContext } from 'react'

interface State {
  vision: boolean
  setVision: (bool: boolean) => void
}

export const CoreStateContext = createContext<State | null>(null)

export function CoreProvider({ children }: { children: React.ReactNode }) {
  const [vision, setVision] = useState(true)

  return (
    <CoreStateContext.Provider value={{ vision, setVision }}>
      {children}
    </CoreStateContext.Provider>
  )
}

export function useCoreState() {
  const state = useContext(CoreStateContext)
  if (!state) throw new Error('Cannot find CoreStateContext')
  return state
}
