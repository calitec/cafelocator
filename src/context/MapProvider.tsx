import {
  useCallback,
  createContext,
  useState,
  useContext,
  useMemo,
  useLayoutEffect,
  useEffect,
} from 'react'
import { IMapDatasProps, IMapDetailProps } from '../types/map'

export interface MapState {
  mapInfo: {
    currentPosition: { lat: number; lng: number }
    mapPosition: { lat: number; lng: number }
    mapDatas: IMapDatasProps[]
    mapDetail: IMapDetailProps
    directions: object
    travel: boolean
    loading: boolean
  }
  setMapInfo: (data: object) => void
  getCurrentLocation: () => void
  onClearDirections: () => void
  onReset: () => void
}

export const initialState = {
  mapInfo: {
    currentPosition: { lat: 0, lng: 0 },
    mapPosition: { lat: 0, lng: 0 },
    mapDatas: null,
    mapDetail: null,
    directions: null,
    travel: false,
    loading: false,
  },
  setMapInfo: null,
  getCurrentLocation: null,
  onClearDirections: null,
  onReset: null,
}

export const MapContext = createContext<MapState | null>(null)

const MapProvider: React.FunctionComponent = ({ children }) => {
  const [mapInfo, setMapInfo] = useState(initialState.mapInfo)
  const { mapDatas, mapPosition, travel, loading } = mapInfo

  // 맵 초기화
  useLayoutEffect(() => {
    if (mapPosition.lat < 1) {
      const { geolocation } = navigator
      if (geolocation) {
        geolocation.getCurrentPosition((position) => {
          const positions = {
            lat: position.coords.latitude,
            lng: position.coords.longitude,
          }
          setMapInfo((prev) => ({
            ...prev,
            currentPosition: positions,
            mapPosition: positions,
          }))
        })
      }
    }
  }, [])

  // 로딩 리셋
  useEffect(() => {
    if (loading && mapDatas !== null)
      setMapInfo((prev) => ({ ...prev, loading: false }))
  }, [mapDatas, loading])

  // 실시간 위치정보 갱신
  useEffect(() => {
    const { geolocation } = navigator
    if (geolocation && travel) {
      const id = geolocation.watchPosition((position) => {
        const positions = {
          lat: position.coords.latitude,
          lng: position.coords.longitude,
        }
        setMapInfo((prev) => ({
          ...prev,
          mapPosition: positions,
          currentPosition: positions,
        }))
      })
      return () => geolocation.clearWatch(id)
    }
  }, [travel])

  // 위치정보 갱신
  const getCurrentLocation = useCallback(() => {
    const { geolocation } = navigator
    setMapInfo((prev) => ({ ...prev, loading: true }))
    if (geolocation) {
      geolocation.getCurrentPosition((position) => {
        const positions = {
          lat: position.coords.latitude,
          lng: position.coords.longitude,
        }
        setMapInfo((prev) => ({
          ...prev,
          mapPosition: positions,
          currentPosition: positions,
          loading: false,
        }))
      })
    }
  }, [])

  // 경로/디테일 리셋
  const onClearDirections = useCallback(() => {
    setMapInfo((prev) => ({
      ...prev,
      mapDetail: null,
      directions: null,
      travel: false,
    }))
  }, [mapInfo])

  // Map 상태 리셋
  const onReset = useCallback(() => {
    setMapInfo((prev) => ({
      ...prev,
      mapDatas: null,
      mapDetail: null,
      directions: null,
      travel: false,
      keyword: '',
    }))
  }, [mapInfo])

  const value = useMemo(
    () => ({
      mapInfo,
      setMapInfo,
      getCurrentLocation,
      onClearDirections,
      onReset,
    }),
    [mapInfo]
  )

  return <MapContext.Provider value={value}>{children}</MapContext.Provider>
}

export function useMapContext() {
  const state = useContext(MapContext)
  if (!state) throw new Error('Cannot find MapContext')
  return state
}

export default MapProvider
