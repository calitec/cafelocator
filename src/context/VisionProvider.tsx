import { useEffect } from 'react'
import { useCallback } from 'react'
import { createContext, useState, useContext } from 'react'
interface State {
  vision: boolean
  setVision: (v: boolean) => void
}

const VisionContext = createContext<State | null>(null)

export function VisionProvider({ children }: { children: React.ReactNode }) {
  const [vision, setVision] = useState(true)

  return (
    <VisionContext.Provider value={{ vision, setVision }}>
      {children}
    </VisionContext.Provider>
  )
}

export function useVisionContext() {
  const state = useContext(VisionContext)
  if (!state) throw new Error('Cannot find VisionContext')
  return state
}
