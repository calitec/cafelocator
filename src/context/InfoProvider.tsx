import * as React from 'react'
import {
  useEffect,
  useCallback,
  createContext,
  useState,
  useContext,
} from 'react'
import { useCoreState } from './CoreProvider'

interface State {
  infoPosition: {
    transY: number
  }
  setInfoPosition: (transY: object) => void
  onTouchStart: (e: React.TouchEvent<HTMLDivElement>) => void
  onTouchMove: (e: React.TouchEvent<HTMLDivElement>) => void
  onTouchEnd: (e: React.TouchEvent<HTMLDivElement>) => void
}

export const InfoStateContext = createContext<State | null>(null)

export function InfoProvider({ children }: { children: React.ReactNode }) {
  const { vision, setVision } = useCoreState()
  const [infoPosition, setInfoPosition] = useState({
    transY: 0,
    touchStart: false,
    touchMove: false,
    touchEnd: false,
  })
  const { transY, touchStart, touchMove, touchEnd } = infoPosition

  useEffect(() => {
    if (vision) {
      setInfoPosition({
        ...infoPosition,
        transY: 0,
        touchStart: false,
        touchMove: false,
        touchEnd: false,
    })
    }
  }, [vision])

  useEffect(()=>{
    if(touchEnd) {
      return () => {
        window.removeEventListener('touchstart', onTouchStart)
        window.removeEventListener('touchmove', onTouchMove)
        window.removeEventListener('touchend', onTouchEnd)
      }
    }
  },[touchEnd])

  const onTouchStart = useCallback(
    (e) => {
      setInfoPosition({
        ...infoPosition,
        transY: 0,
        touchStart: true,
        touchEnd: false,
      })
    },
    [infoPosition]
  )
  const onTouchMove = useCallback(
    (e) => {
      if (!touchStart) return
        const draggingDOMHeight = 34
        const iphoneXSafeArea = 145
        const contentHeight = 350 + draggingDOMHeight
        const mobileScreenHeight = window.screen.availHeight
        const clientY = e.changedTouches[0].clientY
        const device = clientY - (mobileScreenHeight - contentHeight - 13)
        const deviceX =
          clientY - (mobileScreenHeight - contentHeight - iphoneXSafeArea)
        requestAnimationFrame(() => {
          setInfoPosition({
            ...infoPosition,
            transY: mobileScreenHeight == 812 ? deviceX : device,
            touchStart: true,
            touchMove: true,
          })
        })
    },
    [infoPosition]
  )
  const onTouchEnd = useCallback(
    (e) => {
      e.stopPropagation()
      if (touchStart && !touchMove) {
      }
      if (transY > 170 && transY < 350) {
        setVision(false)
        setInfoPosition({
          ...infoPosition,
          transY: 350,
          touchStart: false,
          touchMove: false,
          touchEnd: true,
        })
      } else {
        setVision(true)
        setInfoPosition({
          ...infoPosition,
          transY: 0,
          touchStart: false,
          touchMove: false,
          touchEnd: true,
        })
      }
      e.preventDefault()
    },
    [vision, infoPosition]
  )

  return (
    <InfoStateContext.Provider
      value={{
        infoPosition,
        setInfoPosition,
        onTouchStart,
        onTouchMove,
        onTouchEnd,
      }}
    >
      {children}
    </InfoStateContext.Provider>
  )
}

export function useInfoState() {
  const state = useContext(InfoStateContext)
  if (!state) throw new Error('Cannot find InfoStateContext')
  return state
}
