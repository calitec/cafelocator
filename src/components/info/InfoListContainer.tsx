import * as React from 'react'
import { useMapState } from '../../context/MapProvider'
import haversine from 'haversine'
import useTouch from 'src/lib/hooks/useTouch'
import useSWR from 'swr'
import { fetcher } from 'src/lib/fetcher'
import InfoListPresenter from './InfoListPresenter'
import useGetDatas from 'src/lib/hooks/useGetDatas'
import useGetDetail from 'src/lib/hooks/useGetDetail'

function haversined(mapPosition, value) {
  return (
    Math.round(
      haversine(
        {
          latitude: mapPosition.lat,
          longitude: mapPosition.lng,
        },
        {
          latitude: value?.geometry.location.lat,
          longitude: value?.geometry.location.lng,
        }
      ) * 100
    ) / 100
  )
}
const InfoListContainer: React.FunctionComponent = () => {
  const { infoPosition } = useTouch()
  const { transY } = infoPosition
  const { mapInfo } = useMapState()
  const { mapPosition, keyword } = mapInfo
  const url = useGetDatas()
  const { onClick } = useGetDetail()
  const { data, error } = useSWR(url, fetcher, { suspense: true })

  return (
    <InfoListPresenter
      mapPosition={mapPosition}
      mapDatas={data}
      keyword={keyword}
      transY={transY}
      onClick={onClick}
      haversined={haversined}
    />
  )
}

export default InfoListContainer
