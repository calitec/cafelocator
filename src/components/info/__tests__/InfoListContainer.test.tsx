import { Suspense } from "react";
import { render } from "@testing-library/react";
import InfoListContainer from '../InfoListContainer'
import Loader from "../../common/Loader";
import { initialState, MapStateContext } from "../../../context/MapProvider";

describe('InfoListContainer test', () => {
  it('should load Suspense', () => {
    const { container } = render(
      <Suspense fallback={<Loader />}>
        <MapStateContext.Provider value={initialState}>
          <InfoListContainer />
        </MapStateContext.Provider>
      </Suspense>
    );
    expect(container).toMatchSnapshot();
  });
});