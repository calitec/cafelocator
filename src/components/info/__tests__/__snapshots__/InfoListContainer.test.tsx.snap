// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoListContainer test should load Suspense 1`] = `
<div>
  <div
    css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
  >
    <ul
      class="infoList"
    >
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
        <div
          class="spinner-border text-success"
        />
      </div>
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
        <p />
        <div />
        <div />
      </div>
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
        <p />
        <div />
        <div />
      </div>
      <div
        css="You have tried to stringify object returned from \`css\` function. It isn't supposed to be used directly (e.g. as value of the \`className\` prop), but rather handed to emotion so it can handle it (e.g. as value of \`css\` prop)."
      >
        <p />
        <div />
        <div />
      </div>
    </ul>
  </div>
</div>
`;
